Source/Tools/FEXOfflineCompiler/Main.cpp

```diff
@@ -0,0 +1,284 @@
+    }
+    std::filesystem::rename(FilenameNew.c_str(), Filename.c_str());
+    close(fd);
+    return Filename;
+  }
+}
+
+// Command handler that parses the given code map and generates a code cache for the selected x86 binary.
+// If no binary is selected explicitly, it is inferred from the code map ExecutableFileId block.
+static int GenerateCache(int argc, const char** argv) {
+  optparse::OptionParser Parser {};
+  Parser.add_option("--outdir").set_default(FEX::Config::GetCacheDirectory() + "cache").help("Output directory for generated cache files");
+  Parser.add_option("--fileid").help("Select binary to generate cache for");
+
+  optparse::Values Options = Parser.parse_args(argc, argv);
+  if (Parser.args().size() != 1) {
+    Parser.print_usage();
+    return 1;
+  }
+  const auto CodeMapPath = (std::string)Parser.args()[0];
+
+  std::ifstream Codemap(CodeMapPath.c_str(), std::ios_base::binary);
+  if (!Codemap) {
+    fmt::print("Could not open {}\n", CodeMapPath);
+    return 1;
+  }
+
+  FEXCore::ExecutableFileInfo ProgramName;
+  std::map<FEXCore::ExecutableFileInfo, fextl::set<uintptr_t>> Data;
+  {
+    auto Parsed = FEXCore::CodeMap::ParseCodeMap(Codemap);
+
+    // If an explicit file id is selected, use it.
+    // Otherwise, fall back to an IsExecutable marker (or pick the first entry if there's only one)
+    auto ExplicitFileId = strtoull(((fextl::string)Options.get("fileid")).data(), nullptr, 16);
+    if (ExplicitFileId) {
+      ProgramName.FileId = ExplicitFileId;
+      ProgramName.Filename = Parsed.at(ExplicitFileId).Filename;
+    }
+
+    for (auto& [FileId, Contents] : Parsed) {
+      if (!ExplicitFileId && (Contents.IsExecutable || Parsed.size() == 1)) {
+        ProgramName.FileId = FileId;
+        ProgramName.Filename = Contents.Filename;
+      }
+      Data.emplace(std::piecewise_construct, std::forward_as_tuple(nullptr, FileId, std::move(Contents.Filename)),
+                   std::forward_as_tuple(std::move(Contents.Blocks)));
+    }
+  }
+  if (!ProgramName.FileId) {
+    fmt::print("Cannot generate cache from unsanitized code map {}", CodeMapPath);
+    return 1;
+  }
+
+  for (auto& [File, Blocks] : Data) {
+    if (!Blocks.empty()) {
+      fmt::print("Parsed {} codemap entries for {} ({:016x})\n", Blocks.size(), File.Filename, File.FileId);
+    } else {
+      fmt::print("Found dependency {} ({:016x})\n", File.Filename, File.FileId);
+    }
+  }
+
+  if (!Data.contains(ProgramName)) {
+    throw std::runtime_error(fmt::format("Input code map {} did not contain {} ({:016x})", CodeMapPath, ProgramName.Filename, ProgramName.FileId));
+  }
+
+  fextl::string OutDir(Options.get("outdir"));
+  if (!OutDir.ends_with('/')) {
+    OutDir.push_back('/');
+  }
+  std::filesystem::create_directories(OutDir);
+
+  const auto PortableInfo = FEX::ReadPortabilityInformation();
+  char* envp[] = {nullptr};
+  FEX::Config::LoadConfig("", envp, PortableInfo);
```

Comment:
It's worth making this an option or picking it up from codemaps etc, or if not at least a comment/TODO. Per-game configs should still work with caching ideally. (I realise FEXServer makes this awkward as you need to serialize env vars in a way WoA doesn't need as the compiler always runs under the game context)

Reply:
Oh will FEXServer be invoking OfflineCompiler rather than calling it as a library?